### PR TITLE
Update rocketboard_16 with *most* of its final features

### PR DESCRIPTION
**Pull request by fl3tching101**
## Description

- Added an animated "splash screen" that plays on bootup
- Re-arranged the display's items to be a lot more space conservative and look more appealing
- Changed the default layout to be a bit more useful and have two layers
- Added an "examine" key feature that allows the user to set the pad into a mode that "examines" keys that are pressed and shows the keycode the key is assigned - this mode is toggled on and off with the custom keycode "KC_EXAM"

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/rocketboard_16/custom_font.h
Comment (drashna) [inline keyboards/rocketboard_16/custom_font.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: platforms/chibios/common/ld/STM32F103x8_stm32duino_bootloader.ld
Comment (drashna) [inline platforms/chibios/common/ld/STM32F103x8_stm32duino_bootloader.ld:22]: This change should be reverted.  Or part of a separate PR

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/rocketboard_16/custom_font.h
Comment (fl3tching101) [reply keyboards/rocketboard_16/custom_font.h:15]: Yeah, no problem. Done.

File: platforms/chibios/common/ld/STM32F103x8_stm32duino_bootloader.ld
Comment (fl3tching101) [reply platforms/chibios/common/ld/STM32F103x8_stm32duino_bootloader.ld:22]: Right, so I was kind of afraid of that. I would guess you are aware of the STM32F103C8T6 "feature" where it is supposed to have 64k of flash, but actually comes with 128k. 

This is known to QMK as a whole because the EEPROM emulation on that exact MCU is placed at the top of 128k. 

The reason I have it being committed here is because I am barely over the 64k - 0x2000 limit. I don't think I am actually over 64k, but within 0x2000 anyway. What should I do? This works, I've already tested it on a couple of boards no problem. I can commit it as it's own commit if you want, but since it is technically not a 128k part I wasn't sure what y'all would think.

Comment (drashna) [reply platforms/chibios/common/ld/STM32F103x8_stm32duino_bootloader.ld:22]: No, I wasn't really aware of it, though, I vaguely remember seeing something about it on discord. 

However, We generally try to keep core and keyboard PRs separate. 

I would prefer that it be a separate PR, and either wait on this being merged, or maybe see about using the nano spec, I think it was. Or if a custom ld can be used here? 

But I'll defer to @tzarc on this, I think.

Comment (fl3tching101) [reply platforms/chibios/common/ld/STM32F103x8_stm32duino_bootloader.ld:22]: Great, thanks. If there was some way to override the size from within the keyboard folder that'd be ideal. I'll wait for recommendations from @tzarc.

Comment (sigprof) [reply platforms/chibios/common/ld/STM32F103x8_stm32duino_bootloader.ld:22]: Note that clone chips like CKS32F103C8T6, which are often found on Bluepill boards from AliExpress, don't have the extra 64K flash: #7204. I guess this is not a problem when you are making your own PCB and can control what components are actually used.

In the longer term this should be solved with something like #8179.

Comment (fl3tching101) [reply platforms/chibios/common/ld/STM32F103x8_stm32duino_bootloader.ld:22]: I agree, in the case of clone chips this would be an issue. But to be clear, the EEPROM emulation would also be an issue since it is done on the assumption that 128k is available. Perhaps there could be a separate setting for clones? Or the flash suffix would also work of course. If there is currently no way to make this flash size change in a keyboard specific way (which it sounds like there isn't) then I can go ahead and remove that change and the PR can proceed. The only issue being that I would need to update the README or something to specify that for the time being that file has to be changed before it will build properly.

Comment (fl3tching101) [reply platforms/chibios/common/ld/STM32F103x8_stm32duino_bootloader.ld:22]: I also noticed in the `STM32F103x8_stm32duino_bootloader.ld` file 0x2000 or 8kb of flash is set aside (for EEPROM presumably), isn't that a ton? It looks like from `eeprom_stm32.h` that only 0x800 or 2kb is used. See line 46 and 47. So an additional 6kb of space could be freed up without issue by changing the 0x2000 to 0x800. Correct? If so I have no issue reverting the 128k change and making a second PR to fix this as that should allow plenty of space for my changes. Win-win.

Comment (sigprof) [reply platforms/chibios/common/ld/STM32F103x8_stm32duino_bootloader.ld:22]: `STM32F103x8_stm32duino_bootloader.ld ` actually reserves 8K of flash for the bootloader (currently the bootloader binaries are around 7K), and does not reserve anything for the EEPROM (which is another problem that should really be considered, and it's actually a concern for many other chips, not just F103).

In theory you can add the custom ld file to your board files and specify it in `MCU_LDSCRIPT`, but you would also need to remove `BOOTLOADER = stm32duino` and copy the rest of setup which is normally done by `bootloader.mk` to your `rules.mk`:
```make
OPT_DEFS += -DBOOTLOADER_STM32DUINO
BOARD = STM32_F103_STM32DUINO
STM32_BOOTLOADER_ADDRESS = 0x80000000
DFU_ARGS = -d 1EAF:0003 -a 2 -R
DFU_SUFFIX_ARGS = -v 1EAF -p 0003
```
(You cannot rely on the code from `bootloader.mk`, because it overrides `MCU_LDSCRIPT` unconditionally, so that it would also override the default setting from `quantum/mcu_selection.mk`.)

Obviously such code copying is not really desired, and the proper solution would probably involve something like #8179 with subsequent makefile modifications to select the ld file and other options according to the specified MCU flash size.

**Comment by fl3tching101** [comment]
I believe the latest commit should solve the issue. @sigprof made a good temporary solution at least. Should get it working, hopefully that #8179 change will solve it more permanently.

**Review by drashna — APPROVED** [review]

File: keyboards/rocketboard_16/custom_font.h
Comment (fauxpark) [inline keyboards/rocketboard_16/custom_font.h:17]:
```suggestion
#include "progmem.h"
```

File: keyboards/rocketboard_16/rules.mk
Comment (fauxpark) [inline keyboards/rocketboard_16/rules.mk:13]:
```suggestion
# Bootloader selection
BOOTLOADER = stm32duino

# Custom loader configuration
MCU_LDSCRIPT = STM32F103C8T6_custom
```

File: keyboards/rocketboard_16/keycode_lookup.h
Comment (fauxpark) [inline keyboards/rocketboard_16/keycode_lookup.h:37]:
```suggestion
typedef struct
{
    char key_string[17];
    uint16_t keycode;
} lookup_table_t;

char* translate_keycode_to_string(uint16_t code);

extern lookup_table_t lookup_table[366];
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/rocketboard_16/custom_font.h
Comment (fl3tching101) [reply keyboards/rocketboard_16/custom_font.h]: Fixed this in [e5590bc](https://github.com/qmk/qmk_firmware/pull/12537/commits/e5590bcec542a8a82b2c2bb5f34f332b59885d87)

File: keyboards/rocketboard_16/keycode_lookup.h
Comment (fl3tching101) [reply keyboards/rocketboard_16/keycode_lookup.h]: Fixed this in [e5590bc](https://github.com/qmk/qmk_firmware/pull/12537/commits/e5590bcec542a8a82b2c2bb5f34f332b59885d87)

File: keyboards/rocketboard_16/rules.mk
Comment (fl3tching101) [reply keyboards/rocketboard_16/rules.mk]: This change completely undoes the solution. See @sigprof's explanation [here](https://github.com/qmk/qmk_firmware/pull/12537#discussion_r625119378l). Tested with your suggested changes and it no longer links due to size issue. With @sigprof's code it links just fine.

Comment (drashna) [reply keyboards/rocketboard_16/rules.mk]: Maybe worth putting a comment in the makefile about the issue.

Comment (fauxpark) [reply keyboards/rocketboard_16/rules.mk]: Alternatively, given the particular issue here, `MCU_LDSCRIPT` could be changed to a `?=` in bootloader.mk, so that my changes do work (and potentially a second ldscript in `platforms/chibios/common/ld` for 128kB chips).

Comment (fl3tching101) [reply keyboards/rocketboard_16/rules.mk]: So... is this good to merge then?

Comment (fl3tching101) [reply keyboards/rocketboard_16/rules.mk]: Bump

**Comment by fl3tching101** [comment]
@fauxpark or @drashna any comments? Or is this good to go?

**Comment by drashna** [comment]
Mostly, it's waiting on a second approval.

File: keyboards/rocketboard_16/ld/STM32F103C8T6_custom.ld
Comment (fauxpark) [inline keyboards/rocketboard_16/ld/STM32F103C8T6_custom.ld:1]: No need for this custom ldscript anymore: https://github.com/qmk/qmk_firmware/tree/master/platforms/chibios/boards/STM32_F103_STM32DUINO/ld
The rules.mk stuff will have to stay though, until the issue sigprof mentioned is resolved.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/rocketboard_16/rules.mk
Comment (fauxpark) [inline keyboards/rocketboard_16/rules.mk:8]:
```suggestion
MCU_LDSCRIPT = STM32F103x8
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/rocketboard_16/rules.mk
Comment (sigprof) [reply keyboards/rocketboard_16/rules.mk:8]: I suppose that should be `MCU_LDSCRIPT = STM32F103xB_stm32duino_bootloader` instead?

Comment (fl3tching101) [reply keyboards/rocketboard_16/rules.mk:8]: Yes! Was just trying to figure out what was going on, @sigprof is correct, with `MCU_LDSCRIPT = STM32F103xB_stm32duino_bootloader` it bootloads just fine. Going to do a bit more testing, but then I will commit/push these changes. Thanks!

File: keyboards/rocketboard_16/keycode_lookup.c
Comment (fauxpark) [inline keyboards/rocketboard_16/keycode_lookup.c:59]: You may want to use the keycode names here instead of their actual values; they are (technically) subject to change in the future.

**Review by fauxpark — CHANGES_REQUESTED** [review]
You also have some submodule changes that need rolling back.

**Comment by fl3tching101** [comment]
> You also have some submodule changes that need rolling back.

@fauxpark Huh, that's weird, I see the four submodules that changed. I'm neither sure how I did that nor how to undo that. Do you have the correct branch/tag they should be on so I can check that branch/tag out? Or is there a better way to fix it?

File: keyboards/rocketboard_16/keycode_lookup.c
Comment (fauxpark) [inline keyboards/rocketboard_16/keycode_lookup.c:391]: These are for Wilba's implementation of RGB Matrix, which is mostly exclusive to his boards.
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/rocketboard_16/keycode_lookup.c
Comment (fl3tching101) [reply keyboards/rocketboard_16/keycode_lookup.c]: Removed these and the other set of unused keycodes

**Review by drashna — APPROVED** [review]

**Comment by fl3tching101** [comment]
@fauxpark In your ballpark

**Review by fauxpark — APPROVED** [review]

**Comment by drashna** [comment]
Thanks!

**Pull request merged**
